includes/data-stores/class-wcs-orders-table-subscription-data-store.php

```diff
@@ -0,0 +1,625 @@
+
+				// Dates are set via update_dates() to make sure relationships between dates are validated
+				if ( 0 === strpos( $prop_key, 'schedule' ) ) {
+					$date_type = str_replace( 'schedule_', '', $prop_key );
+
+					if ( 'start' === $date_type && ! $meta_value ) {
+						$meta_value = $subscription->get_date( 'date_created' );
+					}
+
+					$dates_to_set[ $date_type ] = ( false === $meta_value ) ? 0 : $meta_value;
+				} else {
+					$props_to_set[ $prop_key ] = $meta_value;
+				}
+			}
+		}
+
+		$subscription->update_dates( $dates_to_set );
+		$subscription->set_props( $props_to_set );
+	}
+
+	/**
+	 * Helper method that updates post meta based on an order object.
+	 *
+	 * @param \WC_Subscription $subscription Order object.
+	 *
+	 * @return void
+	 */
+	public function update_order_meta( &$subscription ) {
+		$updated_props = [];
+
+		foreach ( $this->get_props_to_update( $subscription, $this->subscription_meta_keys_to_props ) as $meta_key => $prop ) {
+			$meta_value = ( 'schedule_' === substr( $prop, 0, 9 ) ) ? $subscription->get_date( $prop ) : $subscription->{"get_$prop"}( 'edit' );
+
+			// Store as a string of the boolean for backward compatibility (yep, it's gross)
+			if ( 'requires_manual_renewal' === $prop ) {
+				$meta_value = $meta_value ? 'true' : 'false';
+			}
+
+			$subscription->update_meta_data( $meta_key, $meta_value );
+			$updated_props[] = $prop;
+		}
+
+		do_action( 'woocommerce_subscription_object_updated_props', $subscription, $updated_props );
+
+		parent::update_order_meta( $subscription );
+	}
+
+	/**
+	 * Update subscription dates in the database.
+	 * Returns the date properties saved to the database in array format: [ $prop_name => DateTime Object ]
+	 *
+	 * @param \WC_Subscription $subscription
+	 *
+	 * @return array
+	 */
+	public function save_dates( $subscription ) {
+		global $wpdb;
+
+		$saved_dates    = [];
+		$changes        = $subscription->get_changes();
+		$date_meta_keys = [
+			'_schedule_trial_end',
+			'_schedule_next_payment',
+			'_schedule_cancelled',
+			'_schedule_end',
+			'_schedule_payment_retry',
+			'_schedule_start',
+		];
+
+		// Add any custom date types to the date meta keys we need to save.
+		foreach ( wcs_get_subscription_date_types() as $date_type => $date_name ) {
+			if ( 'last_payment' === $date_type ) {
+				continue;
+			}
+
+			$date_meta_key = wcs_get_date_meta_key( $date_type );
+
+			if ( ! in_array( $date_meta_key, $date_meta_keys, true ) ) {
+				$date_meta_keys[] = $date_meta_key;
+			}
+		}
+
+		$date_meta_keys_to_props = array_intersect_key( $this->subscription_meta_keys_to_props, array_flip( $date_meta_keys ) );
+		$subscription_meta_data  = $this->data_store_meta->read_meta( $subscription );
+
+		// Save the changes to scheduled dates
+		foreach ( $this->get_props_to_update( $subscription, $date_meta_keys_to_props ) as $meta_key => $prop ) {
+			$existing_meta_data = array_column( $subscription_meta_data, null, 'meta_key' )[ $meta_key ] ?? false;
```

Comment:
```suggestion
			$existing_meta_data = $subscription_meta_data[ $meta_key ] ?? false;
```

This change is paired with the change above. Previously having the `array_column` inside the foreach would mean it would generate the array in the nicer format each time for each date key.

Reply:
Resolved in https://github.com/Automattic/woocommerce-subscriptions-core/pull/224/commits/c223fe20ec3cc913e3726be35cbaf07fd24d0857